### PR TITLE
Feature 3.6/aql micro optimizations

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Several micro optimizations for AQL:
* AqlValue::slice() can check for externals more efficiently
* RocksDBEdgeIndex creates RocksDB iterator lazily and only when really needed (saves allocations when data can be served from edge cache)
* SubqueryEndExecutor can reuse its already created objects if they are still alive

- [x] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *shell_server_aql*.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9533/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9550/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9571/

**Comment by jsteemann** [comment]
Test failure in IResearchFeature-test is unrelated.

File: arangod/RocksDBEngine/RocksDBEdgeIndex.cpp [hunk 1]
```diff
@@ -305,6 +299,13 @@ class RocksDBEdgeIndexLookupIterator final : public IndexIterator {
 
   void lookupInRocksDB(VPackStringRef fromTo) {
     // Bad case read from RocksDB
+    if (_iterator == nullptr) {
```
Comment (joerg84) [inline arangod/RocksDBEngine/RocksDBEdgeIndex.cpp:302]: Do I assume correctly that every bad read will set _iterator to nullptr?

**Pull request merged**